core/src/main/java/hudson/slaves/NodeProvisioner.java

```diff
@@ -249,7 +246,29 @@ public void run() {
                                 LOGGER.log(Level.SEVERE, "Unexpected uncaught exception encountered while " 
                                         + "processing provisioned slave " + f.displayName, e);
                             } finally {
-                                itr.remove();
+                                while (true) {
+                                    List<PlannedNode> orig = pendingLaunches.get();
+                                    List<PlannedNode> repl = new ArrayList<PlannedNode>(orig);
+                                    // the contract for List.remove(o) is that the first element i where
+                                    // (o==null ? get(i)==null : o.equals(get(i)))
+                                    // is true will be removed from the list
+                                    // since PlannedNode.equals(o) is not final and we cannot assume
+                                    // that subclasses do not override with an equals which does not
+                                    // assure object identity comparison, we need to manually
+                                    // do the removal based on instance identity not equality
+                                    boolean changed = false;
+                                    for (Iterator<PlannedNode> iterator = repl.iterator(); iterator.hasNext(); ) {
```

Comment:
Well ArrayList.remove is going to iterate them all anyway, and this method is not performance critical... after all you don't provision 1,000's of nodes every second... Jenkins will not really be able to handle more than 100 slaves connecting every minute without otherwise drowning in choke-points in the remoting API. I expect maybe tops 20 removes per minute during active provisioning